File: leet_easy_c++/leet_303/leet_303b.cpp
class NumArray {
public:
    NumArray(vector<int> nums) {
        vec.emplace_back(0);
        vec.insert(vec.end(),nums.begin(),nums.end());
        int len = vec.size();
        for(int i=1; i<len; i++) *(vec.begin()+i) += *(vec.begin()+i-1);
    }
    
    int sumRange(int i, int j) {
        return vec[j+1] - vec[i];
    }
private:
    vector<int> vec;
};

/**
 * Your NumArray object will be instantiated and called as such:
 * NumArray obj = new NumArray(nums);
 * int param_1 = obj.sumRange(i,j);
 */

File: leet_easy_java/leet_303/leet_303b.java
class NumArray {
    private int[] num_ary;
    
    public NumArray(int[] nums) {
        num_ary = nums;
        for(int i=1; i<nums.length; i++) num_ary[i] = num_ary[i-1] + nums[i];
    }
    
    public int sumRange(int i, int j) {
        return (i == 0 ? num_ary[j] : num_ary[j]-num_ary[i-1]);
    }
}

/**
 * Your NumArray object will be instantiated and called as such:
 * NumArray obj = new NumArray(nums);
 * int param_1 = obj.sumRange(i,j);
 */